### PR TITLE
[Refactor] Refactor partial selected for  complex type

### PR DESCRIPTION
**Pull request by Smith-Cruise**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

# This is Demo for discuss.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java [hunk 1]
```diff
@@ -38,6 +38,7 @@ public static TaskScheduler create() {
     @Override
     public void executeTasks(TaskContext context) {
         long timeout = context.getOptimizerContext().getSessionVariable().getOptimizerExecuteTimeout();
+        timeout = 50000000;
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java:41]: why do you change it?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java [hunk 2]
```diff
@@ -284,19 +281,20 @@ public ScalarOperator visitSlot(SlotRef node, Context context) {
...
-                }
-
-                columnRefOperator.addUsedSubfieldPos(ImmutableList.copyOf(usedSubFieldPos));
-
-                for (int i = 0; i < node.getUsedStructFieldPos().size(); i++) {
-                    usedSubFieldPos.pop();
-                }
+                returnValue =
+                        SubfieldOperator.build(columnRefOperator, node.getOriginType(), node.getUsedStructFieldPos());
+
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java:286]: if these are not used any more, you can just delete them

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        @Override
+        public Void visit(OptExpression optExpression, Context context) {
+            for (OptExpression opt : optExpression.getInputs()) {
+                opt.getOp().accept(this, opt, context);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalOlapScan(OptExpression optExpression, Context context) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java:179]: You can just visitPhysicalScan

File: fe/fe-core/src/main/java/com/starrocks/catalog/MapType.java [hunk 4]
```diff
@@ -67,6 +67,16 @@ public void setSelectedField(int pos, boolean needSetChildren) {
         }
     }
 
+    public void pruneUnusedFields() {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/catalog/MapType.java:70]: add some comment

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 5]
```diff
@@ -828,6 +830,9 @@ public boolean getUseScanBlockCache() {
     @VarAttr(name = CBO_REORDER_THRESHOLD_USE_EXHAUSTIVE)
     private int cboReorderThresholdUseExhaustive = 6;
 
+    @VarAttr(name = ENABLE_PRUNE_COMPLEX_TYPES)
+    private boolean enablePruneComplexTypes = false;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:834]: It looks like no used

Comment (Smith-Cruise) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:834]: Used in prune complex types

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+
+        public Void visitChild(OptExpression optExpression, Context context) {
+            for (OptExpression opt : optExpression.getInputs()) {
+                opt.getOp().accept(this, opt, context);
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalOlapScan(OptExpression optExpression, Context context) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java:125]: Use visitPhysicalScan is enough

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+        }
+
+        @Override
+        public Void visitPhysicalDistribution(OptExpression optExpression, Context context) {
+            markForNormalOperator(optExpression, context);
+            return visitChild(optExpression, context);
+        }
+
+        @Override
+        public Void visitPhysicalHashJoin(OptExpression optExpression, Context context) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java:226]: You already use visitPhysicalJoin, there is no need to write visitPhysicalMergeJoin and visitPhysicalHashJoin

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/StructTypeTest.java [hunk 8]
```diff
@@ -41,19 +41,77 @@ public static void beforeClass() throws Exception {
...
+    }
+
     @Test
     public void testSelectArrayStruct() throws Exception {
-        String sql = "select c1.a[10].b from test";
+        String sql = "select c1.b[10].a from test";
         String plan = getFragmentPlan(sql);
-        Assert.assertTrue(plan.contains("1:Project\n" +
-                "  |  <slot 4> : 2: c1.a[10].b"));
+        //        Assert.assertTrue(plan.contains("1:Project\n" +
```
Comment (Youngwb) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/StructTypeTest.java:70]: give right arrert

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+                }
+                Type tmpType = type;
+                for (int i = 0; i < usedSubfieldPos.size(); i++) {
+                    // we will always select the ItemType of ArrayType, so we don't mark it and skip it.
+                    while (tmpType.isArrayType()) {
+                        tmpType = ((ArrayType) tmpType).getItemType();
+                    }
+                    int pos = usedSubfieldPos.get(i);
+                    if (i == usedSubfieldPos.size() - 1) {
+                        // last one, select children's all subfields
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java:367]: Why last one will select children's all subfields

**Review by Youngwb — DISMISSED** [review]

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/StructTypeTest.java (hunk 8 above)
Comment (dirtysalt) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/StructTypeTest.java:70]: @Smith-Cruise agree, better not to delete old case, instead change to rigth value.

**Comment by dirtysalt** [comment]
LGTM.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 261 / 311 (83.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/FileTableScanNode.java|0|6|00.00%|[146, 147, 148, 149, 150, 152]|
|:large_blue_circle:|com/starrocks/planner/HudiScanNode.java|0|6|00.00%|[126, 127, 128, 129, 130, 132]|
|:large_blue_circle:|com/starrocks/planner/IcebergScanNode.java|0|6|00.00%|[343, 344, 345, 346, 347, 349]|
|:large_blue_circle:|com/starrocks/catalog/Type.java|0|1|00.00%|[591]|
|:large_blue_circle:|com/starrocks/planner/DeltaLakeScanNode.java|0|6|00.00%|[255, 256, 257, 258, 259, 261]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|2|9|22.22%|[5025, 5026, 5027, 5028, 5029, 5030, 5031]|
|:large_blue_circle:|com/starrocks/connector/parser/trino/AstBuilder.java|2|9|22.22%|[621, 622, 623, 624, 625, 626, 627]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpressionVisitor.java|7|12|58.33%|[130, 134, 138, 142, 154]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|3|4|75.00%|[772]|
|:large_blue_circle:|com/starrocks/planner/HdfsScanNode.java|5|6|83.33%|[152]|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|6|7|85.71%|[78]|
|:large_blue_circle:|com/starrocks/sql/analyzer/ExpressionAnalyzer.java|8|9|88.89%|[253]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/PruneSubfieldsForComplexType.java|179|181|98.90%|[194, 331]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/SubfieldOperator.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/MapType.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/StructType.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/ArrayType.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AstToStringBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/analysis/SlotDescriptor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/SqlToScalarOperatorTranslator.java|2|2|100.00%|[]|

**Review by dirtysalt — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by Smith-Cruise** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by packy92 — APPROVED** [review]

**Pull request merged**